lib/bindings/python/rust/llm/block_manager/vllm/connector/leader/recorder.rs

```diff
@@ -0,0 +1,261 @@
+        let _ = self.connector_leader.update_state_after_alloc(request_id, block_ids, num_external_tokens).unwrap();
+        let runtime = pyo3_async_runtimes::tokio::get_runtime();
+        let _ = runtime.block_on(async move {
+            self.recorder_tx
+                .send(Action::UpdateStateAfterAlloc(input_copy, UpdateStateAfterAllocOutput {}))
+                .await
+        });
+        Ok(())
+    }
+
+    fn build_connector_metadata(
+        &mut self,
+        scheduler_output: SchedulerOutput,
+    ) -> PyResult<Vec<u8>> {
+        let input_copy = BuildConnectorMetaInput {
+            scheduler_output: scheduler_output.clone(),
+        };
+        let output = self.connector_leader.build_connector_metadata(scheduler_output);
+        let output_copy = output.as_ref().unwrap().clone();
+        let runtime = pyo3_async_runtimes::tokio::get_runtime();
+        let _ = runtime.block_on(async move {
+            self.recorder_tx
+                .send(Action::BuildConnectorMeta(input_copy, BuildConnectorMetaOutput {
+                    metadata: output_copy,
+                }))
+                .await
+        });
+        output
+    }
+
+    fn request_finished(&mut self, request_id: String, block_ids: Vec<BlockId>) -> PyResult<bool> {
+        let input_copy = RequestFinishedInput {
+            request_id: request_id.clone(),
+            block_ids: block_ids.clone(),
+        };
+        let output = self.connector_leader.request_finished(request_id, block_ids);
+        let output_copy = output.as_ref().unwrap().clone();
+        let runtime = pyo3_async_runtimes::tokio::get_runtime();
+        let _ = runtime.block_on(async move {
+            self.recorder_tx
+                .send(Action::RequestFinished(input_copy, RequestFinishedOutput {
+                    is_finished: output_copy,
+                }))
+                .await
+        });
+        output
+    }
+
+    fn has_slot(&self, request_id: String) -> bool {
+        let input_copy = HasSlotInput {
+            request_id: request_id.clone(),
+        };
+        let output = self.connector_leader.has_slot(request_id);
+        let output_copy = output.clone();
+        let runtime = pyo3_async_runtimes::tokio::get_runtime();
+        let _ = runtime.block_on(async move {
+            self.recorder_tx
+                .send(Action::HasSlot(input_copy, HasSlotOutput {
+                    result: output_copy,
+                }))
+                .await
+        });
+        output
+    }
+
+    /// Create a new slot for the given request ID.
+    /// This is used to create a new slot for the request.
+    fn create_slot(&mut self, request: KvbmRequest, tokens: Vec<u32>) -> PyResult<()> {
+        let input_copy = CreateSlotInput {
+            request: request.clone(),
+            tokens: tokens.clone(),
+        };
+        let _ = self.connector_leader.create_slot(request, tokens);
+        let runtime = pyo3_async_runtimes::tokio::get_runtime();
```

Comment:
let's use an unbounded channel here so we can use a non-asynchronous send.  this will avoid the runtime + block_on at each layer.